### DISCUSSION TITLE
Security updates for indirect dependencies for pip-compile

### DISCUSSION THREAD

**Discussion by not7cd** [renovatebot/renovate#26947] [Suggest an Idea]
This is a continuation of thread from pip-compile support feature request https://github.com/renovatebot/renovate/discussions/24725#discussioncomment-8222290

In my specific use-case, I wish to create updates for indirect dependencies that can be only found in lock files `requirements.txt`. As of version 37.162 Renovate creates PRs only for dependencies that exist in package files `requirements.in`. For libraries, we will favour `lockFileMaintenance` but for services we prefer small incremental updates. Hence, this feature request.

In draft PR https://github.com/renovatebot/renovate/pull/26871, commit https://github.com/renovatebot/renovate/pull/26871/commits/cffa2d545be59beab284e4854ad8885f1e0b817d I was able to cobble together a solution that is able to create PRs for all indirect dependencies by setting their `depType` to `pip-indirect`, and executing for each `depType==pip-indirect` function `updateLockedDependency` defined by `pip-compile` manager. This is a workaround and is quite limited as in configuration you need to disable updates for all packages, else it will create updates for all indirect packages. It will also silently fail if command errors out in `updateLockedDependency`. That's where `updateArtifacts` has an advantage, as it always creates PR and will include a comment with error message. This is desirable as it presents a place to start for resolving package conflicts.

My draft PR proves that this is possible for `pip-compile` to update single indirect dependency with `--upgrade-package` option, and it can be incorporated into Renovate.

I have created this discussion to explore ways to better incorporate such feature that will not require any additional configuration and work well out-of-the-box. Because what I understand, it will require a bit of refactoring. Hopefully this can begin a support for other package managers.

Here are similar discussions, I'm not aware of any other related to `pip-compile`:
* related to pep621 (poetry) https://github.com/renovatebot/renovate/discussions/14995
* related to Go https://github.com/renovatebot/renovate/discussions/26703
* related to Go https://github.com/renovatebot/renovate/discussions/22505
* related to Go https://github.com/renovatebot/renovate/discussions/22487
* related to Java https://github.com/renovatebot/renovate/discussions/23610

**Comment by viceice** [comment]
`updateLockedDependency` should not call a package manager.

that should be done in `updateArtifacts`

indirect deps should be disabled by default.

**Comment by not7cd** [comment]
I will list functions, that I think, are crucial for such feature to work.

Here, `updateArtifacts` is called.
https://github.com/renovatebot/renovate/blob/ea5e20ebb1fd7108a70800514b28eba92c27f3a5/lib/workers/repository/update/branch/get-updated.ts#L38-L50

Here in `generateUpdate` is decided if something is a `isLockfileUpdate`. This is the only place in code that sets this flag.
https://github.com/renovatebot/renovate/blob/ea5e20ebb1fd7108a70800514b28eba92c27f3a5/lib/workers/repository/process/lookup/generate.ts#L95-L97

Here in `lookupUpdates` range strategy is adapted for security updates. `lookupUpdates` calls `generateUpdate`. This method also ignores certain dependencies.
https://github.com/renovatebot/renovate/blob/ea5e20ebb1fd7108a70800514b28eba92c27f3a5/lib/workers/repository/process/lookup/index.ts#L228-L234

**Reply by viceice** [reply]
I think you need to extract and return indirect deps. set them to disabled by default. then the security stuff can reenable when required. or a user can do with a package rule of cause

**Reply by viceice** [reply]
I think the go manager already does this

**Comment by not7cd** [comment]
What is the meaning of transitive dependency in the context of `npm` and Renovate?

https://docs.renovatebot.com/configuration-options/#transitiveremediation
https://fossa.com/blog/direct-dependencies-vs-transitive-dependencies/
